From 4357d704381c93414d2508e5c8777c24fa19b0a6 Mon Sep 17 00:00:00 2001
From: Bingwu Zhang <xtex@aosc.io>
Date: Sat, 25 Oct 2025 14:00:04 +0800
Subject: [PATCH 041/133] cbe: fix more MIPS register names in inline assembly
X-Developer-Signature: v=1; a=openpgp-sha256; l=2693; i=xtex@astrafall.org;
 h=from:subject; bh=go8ynThjm8dr2fTpvMljrhmZMfmnevrZkdMOnrTdVWQ=;
 b=owGbwMvMwCW2U4Ij7wZL9ETG02pJDJn7cv/HzVx8OkOQfbVBalTkn+Ku5p8rG5IeJ61fV6bax
 bb5rrdNx0QWBjEuBksxRZYiwwZvVp10ftFl5bIwc1iZQIZIizQwMDAwsDDw5SbmlRrpGOmZahvq
 GRrqGOkYMXBxCsBU/zBj+Ct1fZqCwfulx4rnGPTMP3DkZc1j4QLt7tTXX3QTOnRvbrjJyNCn+/n
 r1tvc9o96Lu//rND19v+/XVnGu18qrWDeJtDy9jUTAA==
X-Developer-Key: i=xtex@astrafall.org; a=openpgp;
 fpr=7231804B052C670F15A6771DB918086ED8045B91

(cherry picked from commit a8fea09cd59601bf2eb84643291be196c4ef2a2f)
---
 src/codegen/c.zig | 30 +++++++++++++++++-------------
 1 file changed, 17 insertions(+), 13 deletions(-)

diff --git a/src/codegen/c.zig b/src/codegen/c.zig
index a33a2faf7c26..6c8d321f835b 100644
--- a/src/codegen/c.zig
+++ b/src/codegen/c.zig
@@ -5768,21 +5768,25 @@ fn airAsm(f: *Function, inst: Air.Inst.Index) !CValue {
         switch (aggregate.storage) {
             .elems => |elems| for (elems, 0..) |elem, i| switch (elem) {
                 .bool_true => {
-                    const name = struct_type.structFieldName(i, zcu).toSlice(ip).?;
-                    assert(name.len != 0);
+                    const field_name = struct_type.structFieldName(i, zcu).toSlice(ip).?;
+                    assert(field_name.len != 0);
 
                     const target = &f.object.dg.mod.resolved_target.result;
-                    if (target.cpu.arch.isMIPS() and name[0] == 'r') {
-                        // GCC uses "$N" for register names instead of "rN" used by Zig.
-                        var c_name_buf: [4]u8 = undefined;
-                        const c_name = (&c_name_buf)[0..name.len];
-                        @memcpy(c_name, name);
-                        c_name_buf[0] = '$';
-
-                        try w.print(" {f}", .{fmtStringLiteral(c_name, null)});
-                        (try w.writableArray(1))[0] = ',';
-                        continue;
-                    }
+                    var c_name_buf: [16]u8 = undefined;
+                    const name =
+                        if (target.cpu.arch.isMIPS() and field_name[0] == 'r') name: {
+                            // Convert "rN" to "$N"
+                            const c_name = (&c_name_buf)[0..field_name.len];
+                            @memcpy(c_name, field_name);
+                            c_name_buf[0] = '$';
+                            break :name c_name;
+                        } else if ((target.cpu.arch.isMIPS() and (mem.startsWith(u8, field_name, "fcc") or field_name[0] == 'w')) or
+                        (target.cpu.arch.isMIPS() and field_name[0] == 'f')) name: {
+                            // "$" prefix for FCC, W and F registers
+                            c_name_buf[0] = '$';
+                            @memcpy((&c_name_buf)[1..][0..field_name.len], field_name);
+                            break :name (&c_name_buf)[0 .. 1 + field_name.len];
+                        } else field_name;
 
                     try w.print(" {f}", .{fmtStringLiteral(name, null)});
                     (try w.writableArray(1))[0] = ',';
-- 
2.52.0

